### DISCUSSION TITLE
Struggles with 'level: 2' even and undefined methods/properties

### DISCUSSION THREAD

**Discussion by EliW** [larastan/larastan#1974] [Q&A]
Hey everyone, I'm sure that this is user-error here on my part.  I'm extremely familiar with phpstan, but newer to Laravel ... but, I'm working on a project that was a Laravel 8.x project on PHP 7.4 ... that was only able to pass phpstan level: 1 ... So many issues (real ones) but also problems of phpstan/larastan not seeming to understand the Laravel 'magic'.   

I assumed that mostly the problem was being locked to old verwsions, so through the help of shift I've managed to upgrade the project now to Laravel 11.x, and PHP 8.3. (yaaaay) ... and now can run v.latest versions of larastan.

But I'm really struggling with how to get it to actually understand what methods/properties exist ... when I go just to level: 2, I suddenly have over 500 errors reported, and they are almost all 'undefined property / undefined method' issues.

I tried then using the laravel ide-helper package, and did the `ide-helper:models` command to generate all the property/method attributes.  But I'm still getting issues.   I'm going to give some examples, but please let me know what other details I can provide:

## ID issue w/ factories example:
So we have some factories for testing, that are giving errors like this:
```
 ------ -------------------------------------------------------------------------------------- 
  Line   database/factories/TenantFactory.php                                                  
 ------ -------------------------------------------------------------------------------------- 
  17     Access to an undefined property Illuminate\Database\Eloquent\Model::$id.              
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property  
  18     Access to an undefined property Illuminate\Database\Eloquent\Model::$id.              
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property  
 ------ -------------------------------------------------------------------------------------- 
```

Looking in that code, here's that section:
```php
class TenantFactory extends Factory
{
    public function definition(): array
    {
        return [
            'property_id' => Property::factory()->create()->id,
            'user_id' => User::factory()->create()->id,
            'first_name' => $this->faker->firstName(),
            'last_name' => $this->faker->lastName(),
[...]
```

If you look in the Property or User models, I have the following defined:
```php
 * @property int $id
```

But the problem seems to be that it doesn't realize that User:factory()->create() ... is actually returning a User model?   But just a 'Model'?

## Another 'just a Model' issue:
This time in an API controller:
```
 ------ -------------------------------------------------------------------------------------- 
  Line   app/Http/Controllers/API/V2/ClaimAPIController.php                                    
 ------ -------------------------------------------------------------------------------------- 
  34     Access to an undefined property Illuminate\Database\Eloquent\Model::$id.              
         💡 Learn more: https://phpstan.org/blog/solving-phpstan-access-to-undefined-property  
  41     Call to an undefined method Illuminate\Database\Eloquent\Model::claims().   
```

Looking in that code we find:
```php
/*29:*/            if ($tenant) {
/*30:*/                $policy = $tenant->qbePolicies()->where('id', $id)->first();
/*31:*/
/*32:*/                if ($policy) {
/*33:*/                    $claim = new Claim();
/*34:*/                    $claim->policy_id = $policy->id;
/*35:*/                    $claim->status = 'Pending';
/*36:*/                    $claim->claim_json = json_encode($input);
/*37:*/                    $claim->carrier_claim_number = '';
/*38:*/                    $claim->carrier_claim_feedback = '{}';
/*39:*/                    $claim->save();
/*40:*/
/*41:*/                    $claims = $policy->claims()->orderByDesc('id')->get();
```

So this all seems to stem from $policy just being seen as a `Model` and not actually a `Policy`.   And here's where this gets perhaps more complicated, because the Policy 'situation' here is a polymorphic relation.  There's a polymorphic table/model called `AbstractPolicy`, which in turn has beneath it up to 3 diff types of policies ... (Enrollments, ThirdPartyPolicies, and 'Policy' (which is the 'qbepolicy', poorly named, another story...)

So that call on line 30 ... is calling this method of Tenant:
```php
    public function qbePolicies(): Builder
    {
        return Policy::whereHas('abstractPolicy', function($query) {
            $query->whereHas('tenants', function($query) {
                $query->where('tenants.id', $this->id);
            });
        });
   }
```

So basically, it's looking at the 'bottom' layer of the polymorph, the qbePolicy (aka Policy), and looking 'up' through the polymorph, to see which ones match this tenant.   

That works, and is returning a Builder, based upon 'Policy' ... Back in the original code you see we are calling `->first()` on this, and so should now have whatever the first 'Policy' model was off that builder query, which was all based upon Policy in the first place.

And if I run this code manually, I see the builder come back, and if I do the ->first() I get a `Policy` model.   But larastan/phpstan don't seem to know that this Builder query, based upon Policy ... will return Policy models?

Honestly I think that 'this' problem is one that happens very frequently of all these errors.  Builder queries that don't understand that the models they are getting, are of the type, they are.

(There is also a flipped version of this.   For example, the above code for qbePolicies() can be refactored instead into just this:
`return $this->abstractPolicies()->whereHasMorph('policyable', Policy::class);`

However, this ends up returning not a 'Builder' but a 'BelongsToMany' instead.   I'm not sure if that actually would solve any issues related here, but in fact it seems to make them worse in other areas.  Because now instead of getting a 'Builder' that internally knows it's based upon the Policy model.   It's a BelongsToMany that is still linked to the polymorphic table itself, not the 'correct' underlying representation.  Which in turn has other issues when you try to 'dive through' it even deeper.

## Anyway
Realizing that this is all a mess, and not sure how useful this text has all been or if anyone can even really help.  But I'm hoping.

**Comment by calebdw** [comment]
Yes I had a similar experience in my work codebase and I've gone from level 0 to level 2, and now I'm working on level 3. It's been a lot of work and I've refused to just use a baseline file that hides all the existing problems away.

The problem is that you need to make use of generics so that Larastan understands what models are being returned. Laravel also has a bad habit of typehinting interfaces *and then proceedes to access methods not defined on the interface* but are defined on the runtime implementation---this is very bad in static analysis. If you type hint an interface (or an abstract class like Model), then you are constrained to only accessing properties / calling methods that are defined on that class.


### Ide-helper

The ide-helper does not always play nice with Larastan---I'm not saying that you shouldn't use them together as they solve different problems, but Larastan does not need the ide-helper to function properly.

### Factories

First and foremost, your factories should have the generics defined (Larastan can do some guess work, but in my experience it pays to be strict with your typing given that this is **static** analysis instead of relying on the magic).

I also prefer to hardwire the model/factory up directly instead of relying on the guesswork:

```php
/** @extends Factory<Tenant> */
class TenantFactory extends Factory
{
    protected $model = Tenant::class;

    public function definition(): array
    {
        return [
            'property_id' => Property::factory()->create()->id,
            'user_id' => User::factory()->create()->id,
            'first_name' => $this->faker->firstName(),
            'last_name' => $this->faker->lastName(),
        ];
    }
}

class Tenant extends Model
{
    // ...

    protected function newFactory(): TenantFactory
    {
        return TenantFactory::new();
    }
}
```

You shouldn't have to define any database backed properties on the model using `@property`, these will be detected from your migrations and/or schema dump.

To reduce the boilerplate on model even more, I created my own `HasFactory` generic trait that simplfies this (but you shouldn't have to use this for Larastan:

<details> 

```php
<?php

declare(strict_types=1);

namespace App\Models\Concerns;

use Illuminate\Database\Eloquent\Factories\Factory;
use Illuminate\Database\Eloquent\Factories\HasFactory as BaseHasFactory;
use Illuminate\Database\Eloquent\Model;

/**
 * @template TFactory of Factory
 */
trait HasFactory
{
    // Must add the following property to the model
    // protected static string $factory;

    use BaseHasFactory {
        factory as protected traitFactory;
    }

    /** @return TFactory */
    public static function factory(
        callable|array|int|null $count = null,
        callable|array $state = [],
    ): Factory {
        return static::traitFactory($count, $state);
    }

    /** @return TFactory */
    protected static function newFactory(): Factory
    {
        return static::$factory::new();
    }
}
```

And then it's used as follows:

```diff
<?php

namespace App\DevOps\Models;

+use App\Models\Concerns\HasFactory;
use Database\Factories\DevOps\ServerStatusFactory;
-use Illuminate\Database\Eloquent\Factories\Factory;
-use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class ServerStatus extends Model
{
+  /** @use<ServerStatusFactory> */
    use HasFactory;

-    protected static function newFactory(): Factory
-    {
-        return ServerStatusFactory::new();
-    }
+    protected static string $factory = ServerStatusFactory::class;
}
```

</details>

### Relationships

It's good to define the relationship generics on the models:

```php
<?php

namespace App\Fax\Models;

use App\Models\User;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Fax extends Model
{
    /** @return HasMany<FaxLog> */
    public function logs(): HasMany
    {
        return $this->hasMany(FaxLog::class);
    }

    /** @return BelongsTo<User, $this> */ // or self instead of $this
    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }
}
```

### Queries

I've found it's best to start queries with `::query()`, this directly returns the builder instance instead of proxying calls to the builder.

Since your query is inside a function call, you need to return the proper generics---PHPStan does not delve into functions to determine what type it returns, it uses the function imports (parameters) and outputs (returns) to do this. When it is inside your function it will validate that you are returning what you say you are, but then it uses the function return type elsewhere in your code. I would write your method as:

```php
    /** @return Builder<Policy> */
    public function qbePolicies(): Builder
    {
        return Policy::query()
            ->whereHas('abstractPolicy', function($query) {
                $query->whereRelation('tenants','id', $this->id);
            });
    }
```

### Collections & arrays

You're going to start needing to define the inner generic types, so instead of just returning `Collection` you need to add generics saying it's a `Collection<int, Tenant>` for instance. Similar ideas for array, but arrays can also define object shapes (like 
`array{foo: string}`, instead of just `array<string, string>`).

### Builders

I prefer dedicated builder classes instead of using scopes, this helps larastan better as well as ide completion.

I have my base Builder class:
```php
<?php

declare(strict_types=1);

namespace App\Common\Builders;

use Illuminate\Database\Eloquent\Builder as EloquentBuilder;
use Illuminate\Database\Eloquent\Model;

/**
 * @template TModelClass of Model
 * @extends \Illuminate\Database\Eloquent\Builder<TModelClass>
 */
class Builder extends EloquentBuilder
{
    /** @var TModelClass */
    protected $model;
}
```

Then I create child builders like:
```php
<?php

declare(strict_types=1);

namespace App\Fax\Builders;

use App\Fax\Models\Fax;
use App\Common\Builders\Builder;

/** @extends Builder<Fax> */
class FaxBuilder extends Builder
{
    /** @return $this */
    public function faxNumber(string $faxNumber): static
    {
        return $this->where('fax_number', 'like', "%{$faxNumber}%");
    }

    /** @return $this */
    public function subject(string $subject): static
    {
        return $this->where('subject', 'ilike', "%{$subject}%");
    }
}
```

And then I link it up to the model with a `HasBuilder` trait similar to the `HasFactory`:

```php
<?php

declare(strict_types=1);

namespace App\Models\Concerns;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;

/**
 * @template TBuilder of Builder
 * @mixin Model
 */
trait HasBuilder
{
    // Must add the following property to the model
    // protected static string $builder;

    /** @return TBuilder */
    public static function query()
    {
        /** @var TBuilder needed to override parent type*/
        return parent::query();
    }

    /**
     * @param \Illuminate\Database\Query\Builder $query
     * @return TBuilder
     */
    public function newEloquentBuilder($query)
    {
        return new static::$builder($query);
    }
}

class Fax extends Model
{
    /** @use HasBuilder<FaxBuilder> */
    use HasBuilder;

    protected static string $builder = FaxBuilder::class;
}
```

**Comment by canvural** [comment]
Hi,

As @calebdw brilliantly answered these are errors all related to missing generic annotations. Larastan makes heavy use of generics to understand what is going on.

PHPStan also has an option that warns you if you do not specify generic types for a generic class. But it does not get activated until level 6. Though if you'd like you can enable it manually by adding `checkGenericClassInNonGenericObjectType: true` to your config file. It'll guide you to fix your code, so that Larastan can understand it better.

**Comment by EliW** [comment]
@calebdw @canvural ... Thank you both for your help.  Because of it I've been churning through making changes to the code and taking a chunk out of moving to next levels.

I did have one followup question however.  While putting the generics in everywhere as needed does appear to be working to let larastan know "what is what" ...

One separate issue that I'm dealing with, appears to be that larastan/phpstan doesn't understand that the attributes that are cast as a date type, are actually Carbon\Carbon instances and not just strings?

I'm getting an awful lot of errors (on many models) that follow the pattern:
```
 ------ ---------------------------------------- 
  Line   app/Models/Tenant.php                   
 ------ ---------------------------------------- 
  712    Cannot call method format() on string.  
  713    Cannot call method format() on string.  
  714    Cannot call method format() on string.  
  715    Cannot call method format() on string.  
 ------ ---------------------------------------- 
```

or
```
 ------ ------------------------------------- 
  Line   app/Models/LeaseHistory.php          
 ------ ------------------------------------- 
  93     Cannot call method lte() on string.  
  101    Cannot call method gte() on string.  
  101    Cannot call method lte() on string.  
 ------ ------------------------------------- 
```

In all cases, those are dates and described in the 'casts' as a date or datetime, such that Laravel is returning Carbon instances ... but larastan seems to think they are strings only?
`            'birth_date' => 'date:m/d/Y',`

I've confirmed that in the code where these problems are happening, the variable is in fact a Carbon instance and not a string.

**Reply by calebdw** [reply]
What does your model look like (and the code where the error is generated)? We have passing tests for this:

https://github.com/larastan/larastan/blob/91c3e56a9d08418e7bd4cb2284563cb69c0f8212/tests/Type/data/model-properties.php#L55-L58

**Reply by calebdw** [reply]
It could be the `:m/d/Y` part is causing an issue

**Reply by EliW** [reply]
@calebdw I can try to make a minimal repeatable example ... but of course the real life version right now is messy.   I did test changing the definition to just `date` or just `datetime` but that didn't fix it.    

Without cleaning up a MVP ... here's what the appropriate bits of code causing this look like in one case ... ooh, this is self contained in a single model as well so a pretty good version:

```php
class LeaseHistory extends Model
{
    protected function casts(): array
    {
        return [
            'id' => 'integer',
            'tenant_id' => 'integer',
            'property_id' => 'integer',
            'lease_start' => 'date',
            'lease_end' => 'date',
        ];
    }

    public function status(): string
    {
        if ($this->lease_start) {
            if (! $this->lease_end) {
                $today = Carbon::now()->startOfDay();
                if ($this->lease_start->lte($today)) {
                    return 'Current (Month-to-Month)';
                } else {
                    return 'Future (Month-to-Month)';
                }
       [...]
    }
```

So `lease_start` is declared as a cast to date ... and is in fact a Carbon instance in this code ... but larastan/phpstan is claiming that it's a string.

And in my schema file, here's the table parts, also defined as a date
```sql
CREATE TABLE `lease_histories` (
  `id` int unsigned NOT NULL AUTO_INCREMENT,
  `tenant_id` bigint unsigned NOT NULL,
  `property_id` bigint unsigned DEFAULT NULL,
  `lease_start` date DEFAULT NULL,
  `lease_end` date DEFAULT NULL,

```

**Reply by calebdw** [reply]
The problem is that Larastan doesn't know what your casts are, you need the `@return array` above the casts method:

https://github.com/larastan/larastan/blob/2.x/UPGRADE.md#upgrading-to-296-from-295

**Reply by EliW** [reply]
@calebdw Well, OK then, that fixes that issue ... but leaves me wondering "why"?  For a few reasons:

1. It's stated in multiple places in larastan that it looks at your schema/migrations to understand your model's data.  (and/or it would make sense for it to look at the casts return here)   So why is it relying instead on the annotation here?
2. It seems ... poor form to require literally duplicating the full casts return as an array shape, when it's a fixed array, exactly known return value.  Normally phpstan picks up on this perfectly without annotations, knowing that a fixed return value, is that value always.

So why is this 'duplicate every casts array' step required?

**Reply by EliW** [reply]
FWIW, I at least found a simpler solution to duplicating the array ... tho I'm still confused as to why it's needed, but I tried the following on a whim knowing how phpstan looks at things, and it works without needing to duplicate:
```php
     const CAST = [
         'id' => 'integer',
         'tenant_id' => 'integer',
         'property_id' => 'integer',
         'lease_start' => 'date',
         'lease_end' => 'date',
         'move_in' => 'date',
         'move_out' => 'date',
         'address' => 'string',
         'city' => 'string',
         'state' => 'string',
         'zip_code' => 'string',
         'unit' => 'string',
    ];

    /**
     * @return self::CAST
     */
    protected function casts(): array
    {
        return self::CAST;
    }
```

**Reply by calebdw** [reply]
:shrug: I had a working PR (https://github.com/larastan/larastan/pull/1878) that would not have required duplicating the casts in an annotation, you can follow the discussion there.

**Reply by calebdw** [reply]
> FWIW, I at least found a simpler solution to duplicating the array ... tho I'm still confused as to why it's needed, but I tried the following on a whim knowing how phpstan looks at things, and it works without needing to duplicate:
> 
> ```
>      const CAST = [
>          'id' => 'integer',
>          'tenant_id' => 'integer',
>          'property_id' => 'integer',
>          'lease_start' => 'date',
>          'lease_end' => 'date',
>          'move_in' => 'date',
>          'move_out' => 'date',
>          'address' => 'string',
>          'city' => 'string',
>          'state' => 'string',
>          'zip_code' => 'string',
>          'unit' => 'string',
>     ];
> 
>     /**
>      * @return self::CAST
>      */
>     protected function casts(): array
>     {
>         return self::CAST;
>     }
> ```

Just use the `$casts` property instead of the method:

```php
     protected $casts = [
         'id' => 'integer',
         'tenant_id' => 'integer',
         'property_id' => 'integer',
         'lease_start' => 'date',
         'lease_end' => 'date',
         'move_in' => 'date',
         'move_out' => 'date',
         'address' => 'string',
         'city' => 'string',
         'state' => 'string',
         'zip_code' => 'string',
         'unit' => 'string',
    ];
```

**Reply by EliW** [reply]
> Just use the `$casts` property instead of the method:

Except according to the documentation the `$casts` property was removed in Laravel 11.x ... and the method is supposed to be used now instead.    (IIRC, Shift even did this conversion for me, these may have been properties before that shift ran)

**Reply by calebdw** [reply]
The property has not been removed, it still exists. The process looks like this:
- The `HasAttributes` trait is initialized during the Model construction.
- The initialization process merges the cast method with the property:
  - `$this->casts = array_merge($this->casts, $this->casts());`
- Thereafter, the framework uses the `getCasts()` method which effectively just returns the `$casts` property

**Reply by EliW** [reply]
Odd ... was there a reason it was removed from the docs and the method added then instead?    I'd be concerned that a future version of Laravel might remove the $casts completely.

**Reply by calebdw** [reply]
> I'd be concerned that a future version of Laravel might remove the $casts completely.

I could easily see that happening as well, this seems to be a transition designed not to cause breaking changes and give folks the time to move to the `casts` method, after which the `$casts` property could be removed in the next major release or two. But who knows what the future holds?

I understand your confusion that the `$casts` property has been removed from the Laravel `11.x` documentation and your frustration in having to repeat the casts in the `casts()` method annotation. I disagree with the decision, but that is the direction taken by the powers-that-be.

**Reply by EliW** [reply]
@calebdw Thank you SO MUCH for all your help so far on this.   I got level:2 passing on this codebase.   Going to level 3, I ran into an interesting situation which appears related to https://github.com/larastan/larastan/issues/902

` Property App\Models\Policy::$original_start_date (Carbon\Carbon|null) does not accept string. `

And of course many variations of this.    Now I personally am a huge fan of being more explicit and that I think all these should be handled as explicit Carbon:parse() or Carbon: createFromFormat() .... but this codebase was heavily relying on the 'pass in a string, Laravel/Carbon figures it out'.    So it would be a lot of cut/pasting of Carbon::parse()

Reading that above issue, it looks like this 'should' work.  But isn't for some reason for me.   I have a bunch of these date fields, and they are being cast as such: `'original_start_date' => 'date:m/d/Y'` ... or similar.   Any reason why larastan/phpstan isn't picking up on the fact that these date fields on a model, can have strings passed in as well as Carbon|null?

**Reply by calebdw** [reply]
This is on purpose, at level 3 properties have a readable and a writeable type (and they can differ). Passing a string to these properties and expecting them to be automatically cast to a Carbon instance is inherently unsafe (for example, I can pass `'gooblygook'`).

We've created a `carbon` helper that we use around the codebase (we also solely use CarbonImmutable to reduce bugs, but that's another story):

```php
/**
 * Note: providing a timezone does not convert $date to
 * $timezone if $date already has a timezone.
 */
function carbon(
    string|DateTimeInterface|null $date,
    DateTimeZone|string|null $timezone = null,
): CarbonImmutable {
    return CarbonImmutable::parse($date, $timezone);
}

// later
$policy->original_start_date = carbon('2024-06-24');
```

**Reply by EliW** [reply]
Makes sense ... from reading that other ticket, and looking at other docs/examples I found online, it looked like larastan was supposed to support it.    Looks like I need to go fix all those, or explicitly for now build in an ignore rule to match this just to move past.   But honestly pretty simple to just add something like you did.
